test/syntax.jl

```diff
@@ -3276,3 +3276,7 @@ end
     @test m.Foo.bar === 1
     @test Core.get_binding_type(m.Foo, :bar) == Any
 end
+
+# issue 44723
+demo44723(thunk)::Any = Base.Experimental.@opaque () -> true ? 1 : 2
+@test demo44723(7)() == 1
```

Comment:
The argument `thunk` was not required. It was from the code before I minimize it.

```suggestion
demo44723()::Any = Base.Experimental.@opaque () -> true ? 1 : 2
@test demo44723(7)() == 1
```